### PR TITLE
chore: add functional tests

### LINKED ISSUE #40: add unit tests

### PR DESCRIPTION
**Pull request by dlouzan**
- mimic functional testing from upstream verdaccio
- add code build to travis ci

Fixes #40

### CODE REVIEW

**Comment by dlouzan** [comment]
@juanpicado It would be great if you could take a look. It's basically the same approach as verdaccio does itself for unit + functional tests. We adapted some things, like for example for functional tests we start a small express process that mocks the required gitlab calls.

We haven't implemented e2e at the moment, though I think we would take a different approach than verdaccio in that case: use docker compose to launch both verdaccio and gitlab, plus a third image that actually performed the e2e tests against the started environment.

Maybe we can also use this exercise for searching a common approach to support testing for plugins directly from verdaccio. This adds copies of support code that already exists in verdaccio, but at the moment that code is an internal implementation detail from verdaccio, not directly importable.

File: test/functional/auth/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// @flow
+
+import { CREDENTIALS, WRONG_CREDENTIALS } from '../config.functional';
+import {HTTP_STATUS} from "../../lib/constants";
```
Comment (ercanucan) [inline test/functional/auth/index.js]: minor: style

File: test/functional/basic/index.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// @flow
+
+import whoAmI from './whoami';
+import ping from './ping';
+
+
```
Comment (ercanucan) [inline test/functional/basic/index.js]: minor: consistent white-space usage

File: test/functional/teardown.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+// const chalk = require('chalk');
+module.exports = async () => {
+  // console.log(chalk.green('Teardown Verdaccio Functional'));
```
Comment (ercanucan) [inline test/functional/teardown.js]: why commented out code?

**Comment by ercanucan** [comment]
@dlouzan looks good to me. only minor comments.

**Comment by juanpicado** [comment]
@dlouzan I'll take a look, just give me some days.

File: test/functional/teardown.js (hunk 3 above)
Comment (dlouzan) [reply test/functional/teardown.js]: Because I'm really lazy, I will correct it

**Comment by juanpicado** [comment]
> Maybe we can also use this exercise for searching a common approach to support testing for plugins directly from verdaccio. This adds copies of support code that already exists in verdaccio, but at the moment that code is an internal implementation detail from verdaccio, not directly importable.

@dlouzan one of my objectives was exactly externalizing functional internal implementation for other usages. It's in my queue of tasks.

Currently, we are working on lerna approach that will allow us to create more packages with no pain and my next step is put all this logic out of the box, I'm sure we will find a common approach here.

**Comment by dlouzan** [comment]
Are we going to merge this or should I close it?

**Comment by juanpicado** [comment]
@dlouzan @bufferoverflow it's ok for you guys?

**Comment by juanpicado** [comment]
@dlouzan I tested them, fine for me. 👍

**Comment by bufferoverflow** [comment]
Thank you guys! Awesome!

**Pull request merged**